src/input-states.js

```diff
@@ -0,0 +1,236 @@
+// input-states.js
+
+var _ = require('underscore');
+
+/**
+ * Getter and Setter for InPort states.
+ *
+ * @this is not used
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName name of the input port whose state is being recorded
+ * @param socketId index of the socket or undefined
+ * @param state a State object containing both a data payload and an lm.
+ *
+ * @return getter returns the current input port state state.
+ *         setter returns the current context
+ *
+ * Usage to retrieve state:
+ *  inputState(node, vnid) : {portName: currentState, multiPortName: [currentState]}
+ *  inputState(node, vnid, portName, socketId) : currentState
+ * Usage to store state:
+ *  inputState(node, vnid, {portName: currentState, multiPortName: [currentState]}) : this
+ *  inputState(node, vnid, portName, socketId, state) : this
+ */ 
+module.exports = function(node, vnid, portName, socketId, state) {
+    if (!node.inPorts || !node.outPorts) throw Error("This isn't a Component node");
+    else if (!_.isString(vnid)) throw Error("Invalid vnid: " + vnid);
+    else if (_.isString(portName) && !node.inPorts[portName])
+        throw Error("Invalid portName: " + portName);
+    else if (arguments.length > 4)
+        return setPortStateBySocketId(node, vnid, portName, socketId, state);
+    else if (_.isString(portName))
+        return getPortStateBySocketId(node, vnid, portName, socketId);
+    else if (_.isObject(arguments[2]))
+        return setAllPortStates(node, vnid, arguments[2]);
+    else return getAllPortStates(node, vnid);
+}
+
+/**
+ * Returns a hash of port names to port states or array of states if the port is
+ * addressable/multi. This provides the caller with all portNames, distingishes between
+ * addressable/multi and non-addressable/multi ports, and provides the caller with all their
+ * states that will be useful when calling an updater.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ */
+function getAllPortStates(node, vnid) {
+    var inPorts = _.pick(node.inPorts, isPort);
```

Comment:
noflo has other things on inPorts, but our facade strips them out already. I thought it didn't hurt leaving this in.